File: programmers30_42888/src/main/java/study/problem/Solution.java
package study.problem;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.StringTokenizer;

public class Solution {

	public String[] solution(String[] record) {
		List<Record> records = new ArrayList<>();
		Map<String, String> names = new HashMap<>();
		for (String input : record) {
			StringTokenizer st = new StringTokenizer(input);
			String event = st.nextToken();
			String id = st.nextToken();
			if (st.hasMoreTokens()) {
				String name = st.nextToken();
				names.put(id, name);
			}
			records.add(Record.of(event, id));
		}

		return records.stream()
			.filter(r -> !r.getEvent().equals(Event.CHANGE))
			.map(r -> String.format("%s님이 %s", names.get(r.getId()), r.getEvent()))
			.toArray(String[]::new);
	}
}

class Record {

	private final String id;
	private final Event event;

	public Record(String id, Event event) {
		this.id = id;
		this.event = event;
	}

	public static Record of(String event, String id) {
		return new Record(id, Event.getEvent(event));
	}

	public String getId() {
		return id;
	}

	public Event getEvent() {
		return this.event;
	}
}

enum Event {
	ENTER,
	LEAVE,
	CHANGE;

	public static Event getEvent(String event) {
		for (Event e : Event.values()) {
			if (e.name().equalsIgnoreCase(event)) {
				return e;
			}
		}
		throw new IllegalArgumentException("not matching event");
	}

	@Override
	public String toString() {
		switch (this) {
			case ENTER:
				return "들어왔습니다.";
			case LEAVE:
				return "나갔습니다.";
			default:
				return "";
		}
	}
}

File: programmers30_42888/src/main/kotlin/study/problem/kt/Solution.kt
package study.problem.kt

import java.util.*

class Solution {
    private var nameSpace = mutableMapOf<String, String>()

    fun solution(record: Array<String>): Array<String> {
        val records = mutableListOf<Record>()
        for (input in record) {
            val st = StringTokenizer(input)
            val event = st.nextToken();
            val id = st.nextToken();
            if (st.hasMoreTokens()) {
                val name = st.nextToken();
                nameSpace[id] = name;
            }
            records.add(Record(id, Event.getEvent(event)))
        }
        return records
            .filter { it.event != Event.CHANGE }
            .map { "${nameSpace[it.id]}님이 ${it.event}" }
            .toTypedArray()
    }

    class Record(val id: String, val event: Event)

    enum class Event {
        ENTER,
        LEAVE,
        CHANGE;

        override fun toString(): String {
            return when (this) {
                ENTER -> "들어왔습니다."
                LEAVE -> "나갔습니다."
                else -> ""
            }
        }

        companion object {
            fun getEvent(name: String): Event {
                for (e in values()) {
                    if (e.name.equals(name, ignoreCase = true)) {
                        return e
                    }
                }
                throw IllegalArgumentException("not matching Event")
            }
        }
    }
}